### PR TITLE
lgogdownloader 3.2 (new formula)

### PR DESCRIPTION
**Pull request by claui**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

I’m really surprised to see we haven’t got this already. 😄

### CODE REVIEW

File: Formula/lgogdownloader.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  sha256 "9e60af50c18533884ca317778b291152d5f04d93dfd3dc2579f6f965538b2031"
+
+  depends_on "boost"
+  depends_on "htmlcxx"
+  depends_on "jsoncpp"
+  depends_on "liboauth"
+  depends_on "rhash"
+  depends_on "tinyxml2"
+
+  depends_on "cmake" => :build
```
Comment (fxcoudert) [inline Formula/lgogdownloader.rb]: Move `:build` dependencies above the others

File: Formula/lgogdownloader.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  depends_on "help2man" => :build
+  depends_on "pkg-config" => :build
+
+  def install
+    mkdir "build" do
+      system "cmake", "..",
```
Comment (fxcoudert) [inline Formula/lgogdownloader.rb]: See other formulas for how we indent this

File: Formula/lgogdownloader.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        "-DCMAKE_INSTALL_PREFIX=#{prefix}",
+        "-DCMAKE_BUILD_TYPE=Release"
+
+      system "make", "install"
+    end
+  end
+
+  test do
+    assert_equal "LGOGDownloader #{version}",
+      pipe_output("#{bin}/lgogdownloader --version").chomp
```
Comment (fxcoudert) [inline Formula/lgogdownloader.rb]: We need a test that does a bit more… can we try to login and check that it fails?

**Review by fxcoudert — DISMISSED** [review]

File: Formula/lgogdownloader.rb (hunk 2 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Can you please elaborate? I feel the level of indentation is the same as all the other formulas that use the `mkdir` block (four spaces before `mkdir`, six spaces before `system`).

Example: `aces_container.rb`

```
  def install
    mkdir "build" do
      system "cmake", "..", *std_cmake_args
      system "make", "install"
    end
  end
```

File: Formula/lgogdownloader.rb (hunk 3 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: I’ll look into it.

File: Formula/lgogdownloader.rb (hunk 2 above)
Comment (fxcoudert) [reply Formula/lgogdownloader.rb]: I meant for arguments to `cmake`. We usually fit more that one per line, and align next lines on the first argument.

File: Formula/lgogdownloader.rb (hunk 3 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: The GOG credentials are kept somewhere in the `$XDG_CONFIG_HOME` hierarchy. Those credentials may be correct, incorrect, or missing. How would the test even know?

File: Formula/lgogdownloader.rb (hunk 1 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Done.

File: Formula/lgogdownloader.rb (hunk 2 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Fixed.

File: Formula/lgogdownloader.rb (hunk 3 above)
Comment (fxcoudert) [reply Formula/lgogdownloader.rb]: The test will most often run on CI machines, where there are no credentials. So let's assume that it fails to get correct credentials, and check for that.

Comment (claui) [reply Formula/lgogdownloader.rb]: This is what the manpage says about `brew test`:

> `brew test formula` […] should generally indicate to the user if something is wrong with the installed formula.

What value would a login test add to warrant breaking the user contract and very likely confusing users?

Comment (fxcoudert) [reply Formula/lgogdownloader.rb]: From the [formula cookbook](https://github.com/Homebrew/brew/blob/master/docs/Formula-Cookbook.md):

> We want tests that don't require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests.

Printing version number doesn't exercise much of the tool's installation. We want tests that can be run on CI, and tell us whether a formula got broken — for example missing linkage or else — when updated.

Comment (claui) [reply Formula/lgogdownloader.rb]: Thanks for pointing that out @fxcoudert. I agree with you that having more sophisticated tests would be a good thing.

At the same time, user experience is just as important as that. (Imagine a user has an issue with `lgogdownloader` at login time, then runs `brew test` in an attempt to narrow down the problem. If the program itself is fine, then `brew test` should pass. A test failure would mislead the user into thinking there’s a problem with the formula.)

How about we divert `$XDG_CONFIG_HOME` to something like this:

> `FileUtils.mktemp(retain: true) { pwd }`

for the duration of the test? I feel that this might work reliably for both CI and users.

Comment (fxcoudert) [reply Formula/lgogdownloader.rb]: I agree. Test itself is run in a temporary directory, so redirecting `XDG_CONFIG_HOME` to `pwd` would work fine.

Comment (claui) [reply Formula/lgogdownloader.rb]: Done.

It was a bit of a pain because it required a TTY.
That said, the test should now be as robust as it gets.

File: Formula/lgogdownloader.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  depends_on "boost"
+  depends_on "htmlcxx"
+  depends_on "jsoncpp"
+  depends_on "liboauth"
+  depends_on "rhash"
+  depends_on "tinyxml2"
+
+  def install
+    mkdir "build" do
+      system "cmake", "..", "-DCMAKE_INSTALL_PREFIX=#{prefix}",
```
Comment (DomT4) [inline Formula/lgogdownloader.rb]: What's with not just using `std_cmake_args` here? Does that break something?

File: Formula/lgogdownloader.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    assert_equal "LGOGDownloader #{version}",
+      pipe_output("#{bin}/lgogdownloader --version").chomp
+
+    login_cmdline = %W[#{bin}/lgogdownloader --list --retries 1]
+    test_credentials = {
+      :username => "test@example.com\n",
+      :password => "secret\n",
+    }
+    ENV["XDG_CONFIG_HOME"] = mktemp(:retain => true) { pwd }
+    ohai login_cmdline.join(" ")
```
Comment (dunn) [inline Formula/lgogdownloader.rb]: Was this for debugging?

Comment (claui) [reply Formula/lgogdownloader.rb]: It’s deliberate; its purpose is to show which of the two tests are executing, and in the event of failure, to determine which test has failed.

In the first test, the `pipe_output` method logs its command line as well. It just does it internally.

File: Formula/lgogdownloader.rb (hunk 4 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Thanks @DomT4 for pointing this out. As you suggested, the additional flags from `std_cmake_args` don’t seem to break anything. I have now replaced the entire statement with this:

```
system "cmake", "..", *std_cmake_args
```

File: Formula/lgogdownloader.rb (hunk 5 above)
Comment (ilovezfs) [reply Formula/lgogdownloader.rb]: Does the fork/exec not work here?

Comment (claui) [reply Formula/lgogdownloader.rb]: Nope, it doesn’t because `lgogdownloader` requires a TTY for the username/password input. None of the existing shell helpers seem to support that.

File: Formula/lgogdownloader.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    end
+  end
+
+  test do
+    require "pty"
+
+    assert_equal "LGOGDownloader #{version}",
+      pipe_output("#{bin}/lgogdownloader --version").chomp
+
+    login_cmdline = %W[#{bin}/lgogdownloader --list --retries 1]
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: I'd just inline these with commas

Comment (claui) [reply Formula/lgogdownloader.rb]: Done.

File: Formula/lgogdownloader.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  test do
+    require "pty"
+
+    assert_equal "LGOGDownloader #{version}",
+      pipe_output("#{bin}/lgogdownloader --version").chomp
+
+    login_cmdline =
+      [bin/"lgogdownloader", "--list", "--retries", "1"]
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: and then inline them below instead of using splat

File: Formula/lgogdownloader.rb (hunk 5 above)
Comment (ilovezfs) [reply Formula/lgogdownloader.rb]: I don't think we need this ohai. If it's unclear where it's failing add another assert in between

File: Formula/lgogdownloader.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    test_credentials = {
+      :username => "test@example.com\n",
+      :password => "secret\n",
+    }
+    ENV["XDG_CONFIG_HOME"] = mktemp(:retain => true) { pwd }
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: this seems weird. we're already in a temp directory and HOME is already set

File: Formula/lgogdownloader.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+  depends_on "boost"
+  depends_on "htmlcxx"
+  depends_on "jsoncpp"
+  depends_on "liboauth"
+  depends_on "rhash"
+  depends_on "tinyxml2"
+
+  def install
+    mkdir "build" do
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: did you try an in-tree build?

File: Formula/lgogdownloader.rb (hunk 8 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Can we guarantee we’re in a temp directory? I wouldn’t want test to spit files left and right unexpectedly.

Comment (ilovezfs) [reply Formula/lgogdownloader.rb]: You're in `testpath` which is always a temp directory.

File: Formula/lgogdownloader.rb (hunk 7 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: You mean I should inline the array in both places where it is needed?

File: Formula/lgogdownloader.rb (hunk 8 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Thanks for the pointer!

File: Formula/lgogdownloader.rb (hunk 7 above)
Comment (ilovezfs) [reply Formula/lgogdownloader.rb]: It's only needed in one place if you dump the `ohai`

File: Formula/lgogdownloader.rb (hunk 9 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: I’m not quite sure what you mean. I basically just followed the upstream instruction. Or so I think. Care to elaborate?

Comment (ilovezfs) [reply Formula/lgogdownloader.rb]: meaning
```
    system "cmake", ".", *std_cmake_args
    system "make", "install"
```
and no `mkdir "build" do`.

File: Formula/lgogdownloader.rb (hunk 5 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: We have two tests here, which are being executed in sequence:

```
$ brew test lgogdownloader
Testing lgogdownloader
==> /usr/local/Cellar/lgogdownloader/3.2/bin/lgogdownloader --version
==> /usr/local/Cellar/lgogdownloader/3.2/bin/lgogdownloader --list --retries 1
```

The first `ohai` happens implicitly inside `pipe_output`.
The second `ohai` is the explicit one.
If I leave it out, the test suggests that the `--version` part is failing. That’s confusing to the end user. Hence, the manual `ohai`.

How would an additional assert in between help with that?

File: Formula/lgogdownloader.rb (hunk 7 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: I feel there is a quite strong case for the `ohai` here, see my explanation. I’ll gladly lose it though if we find another way that doesn’t compromise the UX of `brew test`.

File: Formula/lgogdownloader.rb (hunk 8 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: … and fixed. ☺️

File: Formula/lgogdownloader.rb (hunk 9 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Yay, your suggestion works just as well! Fixed.

File: Formula/lgogdownloader.rb (hunk 7 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Thinking about it, maybe the first test isn’t necessary after all?
If I delete that, the `ohai` can go, too!

Comment (ilovezfs) [reply Formula/lgogdownloader.rb]: The version test?

Comment (claui) [reply Formula/lgogdownloader.rb]: Yes, I mean the `--version` test. It’s just a nice to have after all; the functional test covers more.

File: Formula/lgogdownloader.rb [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  test do
+    require "pty"
+
+    test_credentials = {
+      :username => "test@example.com\n",
+      :password => "secret\n",
+    }
+    ENV["XDG_CONFIG_HOME"] = pwd
+    reader, writer = PTY.spawn(
+      bin/"lgogdownloader", "--list", "--retries", "1"
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: looks like this fits on one line without going passed the eightieth column.
```
    reader, writer = PTY.spawn(bin/"lgogdownloader", "--list", "--retries", "1")
```

File: Formula/lgogdownloader.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  test do
+    require "pty"
+
+    test_credentials = {
+      :username => "test@example.com\n",
+      :password => "secret\n",
+    }
+    ENV["XDG_CONFIG_HOME"] = pwd
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: I'd just do `= testpath`

File: Formula/lgogdownloader.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    reader, writer = PTY.spawn(
+      bin/"lgogdownloader", "--list", "--retries", "1"
+    )
+    writer.write(test_credentials.values.join(""))
+    writer.close
+    assert_equal("HTTP: Login failed",
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: should all fit on one line and no parentheses needed

File: Formula/lgogdownloader.rb (hunk 10 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Only if you insist. I have a vow to stay within a 68 CPL limit for readability. If I break that, a kitten dies!

File: Formula/lgogdownloader.rb (hunk 11 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: That’s more expressive! Thanks for the pointer.

File: Formula/lgogdownloader.rb [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+  def install
+    system "cmake", ".", *std_cmake_args
+    system "make", "install"
+  end
+
+  test do
+    require "pty"
+
+    test_credentials = {
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: credentials = <<-EOS.undent
      test@example.com
      secret
    EOS

Comment (ilovezfs) [reply Formula/lgogdownloader.rb]: actually we can probably just inline below

File: Formula/lgogdownloader.rb [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+      :username => "test@example.com\n",
+      :password => "secret\n",
+    }
+    ENV["XDG_CONFIG_HOME"] = pwd
+    reader, writer = PTY.spawn(
+      bin/"lgogdownloader", "--list", "--retries", "1"
+    )
+    writer.write(test_credentials.values.join(""))
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]:
```
    writer.write <<-EOS.undent
      test@example.com
      secret
    EOS
```

File: Formula/lgogdownloader.rb (hunk 13 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Good idea. I was reluctant to do without the keys at first. But yours is actually more readable!

File: Formula/lgogdownloader.rb [hunk 15]
```diff
@@ -0,0 +1,40 @@
+class Lgogdownloader < Formula
+  desc "Open source downloader for GOG.com games"
+  homepage "https://sites.google.com/site/gogdownloader"
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: terminating slash `https://sites.google.com/site/gogdownloader/`

File: Formula/lgogdownloader.rb (hunk 14 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Even better 👍

File: Formula/lgogdownloader.rb (hunk 10 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: My bad. I did it and it’s still below 68 😉

File: Formula/lgogdownloader.rb (hunk 11 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Done.

File: Formula/lgogdownloader.rb (hunk 12 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Fixed, thanks.

File: Formula/lgogdownloader.rb (hunk 14 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Done.

File: Formula/lgogdownloader.rb [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+  def install
+    system "cmake", ".", *std_cmake_args
+    system "make", "install"
+  end
+
+  test do
+    require "pty"
+
+    ENV["XDG_CONFIG_HOME"] = testpath
+    reader, writer =
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: I think you can join lines 27 and 28

File: Formula/lgogdownloader.rb (hunk 15 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Fixed, thanks.

However, I feel we should really introduce a `style` rule (or `audit` rule? Can’t tell apart why we have two separate style checks) to enforce that if it’s important.

Comment (ilovezfs) [reply Formula/lgogdownloader.rb]: There was a proposal for one but it made me break out it hives since it seemed like too much pedantry. But I don't mind annoying you on a new formula PR about it to save @vszakats the trouble.

Comment (claui) [reply Formula/lgogdownloader.rb]: Fair enough 😅

File: Formula/lgogdownloader.rb [hunk 17]
```diff
@@ -0,0 +1,37 @@
+class Lgogdownloader < Formula
+  desc "Open source downloader for GOG.com games"
+  homepage "https://sites.google.com/site/gogdownloader/"
+  url "https://sites.google.com/site/gogdownloader/lgogdownloader-3.2.tar.gz"
+  sha256 "9e60af50c18533884ca317778b291152d5f04d93dfd3dc2579f6f965538b2031"
+
+  depends_on "cmake" => :build
+  depends_on "help2man" => :build
+  depends_on "pkg-config" => :build
+
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: no blank line here

File: Formula/lgogdownloader.rb [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+  def install
+    system "cmake", ".", *std_cmake_args
+    system "make", "install"
+  end
+
+  test do
+    require "pty"
+
+    ENV["XDG_CONFIG_HOME"] = testpath
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: I have a stupid question. What happens if we don't set this?

Comment (claui) [reply Formula/lgogdownloader.rb]: Not stupid at all! See [the relevant discussion](https://github.com/Homebrew/homebrew-core/pull/19189#discussion_r144208451).

tl;dr If we don’t set this, the CI will run fine but `brew test` will inexplicably fail for most end users.

File: Formula/lgogdownloader.rb (hunk 16 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: We’d hit _exactly_ 80 chars then. What is this, a black site? Can I see the water thingy again? 😉

File: Formula/lgogdownloader.rb (hunk 17 above)
Comment (claui) [reply Formula/lgogdownloader.rb]: Done.

File: Formula/lgogdownloader.rb (hunk 16 above)
Comment (ilovezfs) [reply Formula/lgogdownloader.rb]: 80 is ok. 81 is torture time.

Comment (claui) [reply Formula/lgogdownloader.rb]: Harsh but fair. Changed.

File: Formula/lgogdownloader.rb [hunk 19]
```diff
@@ -0,0 +1,35 @@
+class Lgogdownloader < Formula
+  desc "Open source downloader for GOG.com games"
```
Comment (ilovezfs) [inline Formula/lgogdownloader.rb]: I think we know it's open source …

Comment (claui) [reply Formula/lgogdownloader.rb]: Agreed and fixed. (It’s almost as if you were slowly running out of nits …)

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Thanks for the :new: formula @claui. Disclaimer: No kittens have been harmed in the production of this formula or any other to the best of our knowledge.
